File: 0144.BinaryTreePreorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # Iteratively with stack
        ans = []
        stack = []
        node = root
        while stack or node:
            if node:
                stack.append(node)
                ans.append(node.val)
                node = node.left
            else:
                node = stack.pop()
                node = node.right
        return ans
        '''
        # Recursive
        if not root:
            return []
        #return [root.val] + self.preorderTraversal(root.left) + self.preorderTraversal(root.right)
        ans = []
        ans.append(root.val)
        ans += self.preorderTraversal(root.left)
        ans += self.preorderTraversal(root.right)
        return ans
        '''

File: 0000-0999/0144.BinaryTreePreorderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func preorderTraversal(root *TreeNode) []int {
    stack := []*TreeNode{}
    ans := []int{}
    for root != nil || len(stack) > 0 {
        if root != nil {
            ans = append(ans, root.Val)
            stack = append(stack, root)
            root = root.Left
        } else {
            root = stack[len(stack)-1].Right
            stack = stack[:len(stack)-1]
        }
    }
    return ans
}